pkg/finality-grandpa/weights.go

```diff
@@ -0,0 +1,24 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"fmt"
+	"math"
+	"math/big"
+)
+
+type VoteWeight uint64
+
+type VoterWeight uint64
+
+func (vw *VoterWeight) checkedAdd(add VoterWeight) (err error) {
+	sum := new(big.Int).SetUint64(uint64(*vw))
+	sum.Add(sum, new(big.Int).SetUint64(uint64(add)))
+	if sum.Cmp(new(big.Int).SetUint64(uint64(math.MaxUint64))) > 0 {
+		return fmt.Errorf("VoterWeight overflow for CheckedAdd")
```

Comment:
sentinel error?

Reply:
so I can check with errors.Is/As?

Reply:
- yes, to be able to be checked with errors.Is/As
- feels more readable
- you can reuse and wrap the error etc.

Reply:
From what I've read a sentinel error is almost always public and adds to your public API.  This error in particular is actually not returned when calling `NewVoterSet`, rather you end up getting a `*VoterSet(nil)` if the `checkedAdd` fails.  Whether that's correct or not is another question, but this is the same behaviour as the reference implementation.  I also subscribe to returning opaque errors most of the time ensuring that the caller of this function isn't checking for a specific type of error.  Dave Cheney talks about this in his [blog](https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully), and he seems well versed on the subject given he wrote the `pkg/errors` package.  I usually use private sentinel errors if there are multiple places in the code that return the same error.  I find it quite nice to be able to search for the error and find the exact line that returns it, rather than determining it's a private sentinel error, and then having to find the single reference of that private sentinel error.

Reply:
I would say I tend to agree with Tim here. I really like sentinal errors but more in cases where the error is being used in many places or represents a certain case that is being checked.